client/actions/auth.js

```diff
@@ -49,39 +59,103 @@ export function logout() {
 
 export function loadCredentials() {
   return (dispatch) => {
-    const token = sessionStorage.getItem('sso-dashboard:apiToken');
-    if (token || window.location.hash) {
-      let apiToken = token;
+    const apiToken = sessionStorage.getItem('sso-dashboard:apiToken');
+    const tokenExpire = sessionStorage.getItem('sso-dashboard:tokenExpire');
+    const accessToken = sessionStorage.getItem('sso-dashboard:accessToken');
+    const userProfile = sessionStorage.getItem('sso-dashboard:userProfile');
 
-      const hash = auth0.parseHash(window.location.hash);
-      if (hash && hash.idToken) {
-        apiToken = hash.idToken;
+    if (apiToken) {
+      const decodedToken = jwtDecode(apiToken);
+      if (isExpired(decodedToken)) {
+        return;
       }
 
-      if (apiToken) {
-        const decodedToken = jwtDecode(apiToken);
-        if (isExpired(decodedToken)) {
-          return;
+      axios.defaults.headers.common.Authorization = `Bearer ${apiToken}`;
+
+      dispatch({
+        type: constants.LOADED_TOKEN,
+        payload: {
+          token: apiToken
         }
+      });
 
-        axios.defaults.headers.common.Authorization = `Bearer ${apiToken}`;
+      dispatch({
+        type: constants.LOGIN_SUCCESS,
+        payload: {
+          token: apiToken,
+          decodedToken,
+          user: decodedToken,
+          issuer: url.parse(decodedToken.iss).hostname
+        }
+      });
+    }
 
-        dispatch({
-          type: constants.LOADED_TOKEN,
-          payload: {
-            token: apiToken
-          }
-        });
+    if (accessToken && !tokenExpired(tokenExpire)) {
+      axios.defaults.headers.common.Authorization = `Bearer ${accessToken}`;
+
+      let user;
 
-        dispatch({
-          type: constants.LOGIN_SUCCESS,
-          payload: {
-            token: apiToken,
-            decodedToken,
-            user: decodedToken
+      try {
+        user = JSON.parse(userProfile);
+      } catch (e) {
+        user = {};
+      }
+
+      dispatch({
+        type: constants.LOADED_TOKEN,
+        payload: {
+          token: accessToken
+        }
+      });
+
+      dispatch({
+        type: constants.LOGIN_SUCCESS,
+        payload: {
+          token: accessToken,
+          decodedToken: user,
+          user: user,
+          issuer: user.email
+        }
+      });
+    }
+
+    if (window.location.hash) {
+      webAuth.parseHash(window.location.hash, function(parseErr, hash) {
+        if (parseErr) {
+          return console.log(parseErr);
+        }
+
+        webAuth.client.userInfo(hash.accessToken, function(infoErr, user) {
+          if (infoErr) {
+            return console.log(infoErr);
           }
+
+          sessionStorage.setItem('sso-dashboard:tokenExpire', new Date().setSeconds(hash.expiresIn));
```

Comment:
Instead of using the exact date, you should do something like `date - 10min` to handle clock skew